core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala

```diff
@@ -16,318 +16,182 @@
+          newStart -= nPadBases
 
-          padSeq ++= refGenomeContext.getBases()
+          nTrimmedBasesBeforeNextPadding = 0
 
         } else {
-          nTrimmedBasesBeforeNextPadding -= 1
-        }
 
-        alleles = alleles.map { a =>
-          Allele.create(
-            padSeq ++ a
-              .getBaseString()
-              .dropRight(nTrimmedBasesBeforeNextPadding + 1)
-              .getBytes(),
-            a.isReference
-          )
-        }
+          nTrimmedBasesBeforeNextPadding += 1
 
-        trimSize += nTrimmedBasesBeforeNextPadding + 1
+        }
 
-        newStart -= nPadBases
+        firstBaseFromRightInRefAllele = allAlleles(0)(
+          allAlleles(0).length - nTrimmedBasesBeforeNextPadding - 1
+        )
+      }
 
-        nTrimmedBasesBeforeNextPadding = 0
+      // trim from left
+      var nLeftTrimBases = 0
+      var firstBaseFromLeftInRefAllele = allAlleles(0)(nLeftTrimBases)
+      val minAlleleLength = allAlleles.map(_.length).min
 
-      } else {
+      while (nLeftTrimBases < minAlleleLength - nTrimmedBasesBeforeNextPadding - 1
+        && allAlleles.forall(_(nLeftTrimBases) == firstBaseFromLeftInRefAllele)) {
 
-        nTrimmedBasesBeforeNextPadding += 1
+        nLeftTrimBases += 1
 
+        firstBaseFromLeftInRefAllele = allAlleles(0)(nLeftTrimBases)
       }
 
-      firstAlleleBaseFromRight = alleles(0).getBases()(
-        alleles(0).length
-        - nTrimmedBasesBeforeNextPadding - 1
-      )
-    }
-
-    // trim from left
-    var nLeftTrimBases = 0
-    var firstAlleleBaseFromLeft = alleles(0).getBases()(nLeftTrimBases)
-    val minAlleleLength = alleles.map(_.length).min
-
-    while (nLeftTrimBases < minAlleleLength - nTrimmedBasesBeforeNextPadding - 1
-      && alleles.forall(_.getBases()(nLeftTrimBases) == firstAlleleBaseFromLeft)) {
+      allAlleles = allAlleles.map { a =>
+        a.drop(nLeftTrimBases)
+          .dropRight(nTrimmedBasesBeforeNextPadding)
+      }
 
-      nLeftTrimBases += 1
+      trimSize += nTrimmedBasesBeforeNextPadding
 
-      firstAlleleBaseFromLeft = alleles(0).getBases()(nLeftTrimBases)
-    }
-
-    alleles = alleles.map { a =>
-      Allele.create(
-        a.getBaseString()
-          .drop(nLeftTrimBases)
-          .dropRight(nTrimmedBasesBeforeNextPadding)
-          .getBytes(),
-        a.isReference
-      )
+      newStart += nLeftTrimBases
 
+      flag = if (trimSize == 0 && nLeftTrimBases == 0) {
+        FLAG_UNCHANGED
+      } else {
+        FLAG_CHANGED
+      }
     }
 
-    trimSize += nTrimmedBasesBeforeNextPadding
-
-    AlleleBlock(
-      alleles,
-      newStart + nLeftTrimBases,
-      unalignedAlleleBlock.end - trimSize
+    InternalRow(
+      newStart,
+      end - trimSize,
+      UTF8String.fromString(allAlleles(0)),
+      ArrayData.toArrayData(allAlleles.tail.map(UTF8String.fromString(_))),
+      UTF8String.fromString(flag)
     )
 
   }
 
   private val PAD_WINDOW_SIZE = 100
 
+  val FLAG_ERROR = "Error"
```

Comment:
We should be consistent about how we report errors. In liftover, I think that we return an error message in the case of failure. If the error message is null, the operation was successful.

Reply:
Whether the variant was changed or not should be communicated with a boolean

Reply:
Also, what's in the rest of the result fields when normalization fails?

Reply:
Changed to what you mention. Have the results to be null in case of error.